src/codegen/generators/expression_generator.rs

```diff
@@ -319,6 +273,64 @@ impl<'a, 'b> ExpressionCodeGenerator<'a, 'b> {
         }
     }
 
+    pub fn generate_direct_access_index(
+        &self,
+        access: &DirectAccessType,
+        index: &AstStatement,
+        access_type: &DataTypeInformation,
+        target_type: &DataType,
+    ) -> Result<IntValue<'a>, CompileError> {
+        let rhs = match index {
+            AstStatement::LiteralInteger { value, .. } => {
```

Comment:
have you tried to treat the LiteralInteger and Reference the same?
so basically you could just not care about what Ast-Statement this is, you could call generate
on it since you're doing an intvalue-check against the generated value anyway

Reply:
I probably was avoiding letting llvm handle that but you are probably right, if it's an int llvm will just optimize the multiplication out